lecture-2/Exercises/faisal-sol.md

```diff
@@ -0,0 +1,101 @@
+In summary, that annonymous function has global scope's variable named `btnNum`, which is equal to 3 as the loop is completed and also the function is invoked after the completion of loop. So we can get the right output by following some ways:-
+##### Change `var` into `let`:-
+As we know `let` is hoisted in block element, so we can change the `var` into `let`.  So it will bind to the function's block and give right output. Here is the code:
+```js 
+const prizes = ['A Unicorn!', 'A Hug!', 'Fresh Laundry!'];
+for (let btnNum = 0; btnNum < prizes.length; btnNum++) {
+    document.getElementById(`btn-${btnNum}`).onclick = ()=>{
+        alert(prizes[btnNum]);
+    };
+}
+```
+##### Using `closure`:-
+Here we can use `closure` concept. We can give a closure to that annonymous function which has a parameter and will store the `butNum` of each iteration by invoking inside the block.
+```js
+for (var btnNum = 0; btnNum < prizes.length; btnNum++) {
+  function message(btnNum){
+        document.getElementById(`btn-${btnNum}`).onclick = ()=>{
+            alert(prizes[btnNum]);
+        };
+    }
+    message(btnNum);
+}
+```
+##### Bind a function from button's DOM event `onclick` attribute:-
+Here we can use an attribute of a button `onclick` which will invoke a function named `message` in the script. Here is the code:-
+```html
+<button id="btn-0" onclick="message(0)">Button 1</button>
+<button id="btn-1" onclick="message(1)">Button 2</button>
+<button id="btn-2" onclick="message(2)">Button 3</button>
+
+<script type="text/javascript">
+  const prizes = ['A Unicorn!', 'A Hug!', 'Fresh Laundry!'];
+  function message(btnNum){
+    alert(prizes[btnNum]);
+  }
+</script>
+```
+
+## _Solution of Ex-2:-_
+The right output don't occur because `i` is `var` here. So it will hoisted in the `assignTorpedo`'s functional scope by initializing `undefined`. As loop continues and the function `torpedoAssignment` will store the reference of that variable `i` in it's closure, so in it's closure it will store `i` is equal to `6` after the loop ends. So we can fix this by following these:
+##### 1.Changing `var` to `let`:-
+```js
+for (let i = 0; i < passengerArray.length; i++) {
+    if (passengerArray[i] == name) {
+      torpedoAssignment = function () {
+        console.log('name: ', name);
+        console.log('#: ', i+1);
+      }
+    }
+}
+```
+##### 2. Using `break`:-
+```js
+for (var i = 0; i < passengerArray.length; i++) {
+    if (passengerArray[i] == name) {
+      torpedoAssignment = function () {
+        console.log('name: ', name);
+        console.log('#: ', i+1);
+      }
+      break;
+    }
+}
+```
+##### 3.Using IIFE:-
+```js
+for (var i = 0; i < passengerArray.length; i++) {
+    if (passengerArray[i] == name) {
+      (function () {
+        console.log('name: ', name);
+        console.log('#: ', i+1);
+      })();
```

Comment:
You need to pass the current value of `i` when invoking so that the variable exists in `Closure`

```js
(function (i) {
    console.log('name: ', name);
    console.log('#: ', i+1);
 })(i);
```

Reply:
yes, here you are printing the output directly instead of calling later. e.g.

```js
var subPassengers = ["Luke", "Leia", "Han", "Alice", "Yoda", "Boba"];
var giveAssignment = assignTorpedo("Alice", subPassengers);

giveAssignment(); // this line is missing
```

Reply:
yes because I don't need that calling, that's why I have removed that line